File: scripts/validate.py
#!/usr/bin/env python3
"""Validate repository-level Agent Skills invariants without dependencies."""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
ALLOWED_FIELDS = {
    "name",
    "description",
    "license",
    "compatibility",
    "metadata",
    "allowed-tools",
}
NAME_PATTERN = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
LINK_PATTERN = re.compile(r"(?<!!)\[[^]]*]\(([^)]+)\)")
TABLE_SKILL_PATTERN = re.compile(r"^\| `([^`]+)` \|", re.MULTILINE)


def frontmatter(path: Path) -> tuple[dict[str, str], list[str]]:
    text = path.read_text(encoding="utf-8")
    lines = text.splitlines()
    errors: list[str] = []
    if not lines or lines[0] != "---":
        return {}, ["missing opening frontmatter delimiter"]
    try:
        end = lines.index("---", 1)
    except ValueError:
        return {}, ["missing closing frontmatter delimiter"]

    fields: dict[str, str] = {}
    for line in lines[1:end]:
        if not line or line.lstrip().startswith("#") or line[0].isspace():
            continue
        match = re.match(r"^([a-z][a-z0-9-]*):(?:\s*(.*))?$", line)
        if not match:
            errors.append(f"invalid top-level frontmatter line: {line!r}")
            continue
        key, value = match.groups()
        fields[key] = value or ""
    return fields, errors


def validate_skill(skill_dir: Path) -> list[str]:
    path = skill_dir / "SKILL.md"
    fields, errors = frontmatter(path)
    unknown = sorted(set(fields) - ALLOWED_FIELDS)
    if unknown:
        errors.append(f"unsupported frontmatter fields: {', '.join(unknown)}")

    name = fields.get("name", "").strip(" '\"")
    if not name:
        errors.append("missing non-empty name")
    elif not NAME_PATTERN.fullmatch(name):
        errors.append(f"invalid skill name: {name!r}")
    elif name != skill_dir.name:
        errors.append(f"name {name!r} does not match directory {skill_dir.name!r}")

    if "description" not in fields:
        errors.append("missing description")

    text = path.read_text(encoding="utf-8")
    for target in LINK_PATTERN.findall(text):
        target = target.split("#", 1)[0]
        if not target or "://" in target or target.startswith(("#", "mailto:")):
            continue
        if target.startswith("/"):
            errors.append(f"absolute Markdown link is not portable: {target}")
        elif not (skill_dir / target).resolve().exists():
            errors.append(f"broken relative Markdown link: {target}")
    return errors


def load_json(path: Path) -> dict:
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as error:
        raise ValueError(f"{path.relative_to(ROOT)}: {error}") from error


def main() -> int:
    failures: list[str] = []
    skill_dirs = sorted(path.parent for path in ROOT.glob("*/SKILL.md"))
    skill_names = {path.name for path in skill_dirs}
    if not skill_dirs:
        failures.append("no top-level skills found")

    for skill_dir in skill_dirs:
        for error in validate_skill(skill_dir):
            failures.append(f"{skill_dir.relative_to(ROOT)}: {error}")

    try:
        catalog = load_json(ROOT / "skills.sh.json")
        catalog_names = [
            name
            for grouping in catalog.get("groupings", [])
            for name in grouping.get("skills", [])
        ]
        if len(catalog_names) != len(set(catalog_names)):
            failures.append("skills.sh.json: a skill appears in more than one grouping")
        missing = sorted(skill_names - set(catalog_names))
        extra = sorted(set(catalog_names) - skill_names)
        if missing:
            failures.append(f"skills.sh.json: missing skills: {', '.join(missing)}")
        if extra:
            failures.append(f"skills.sh.json: unknown skills: {', '.join(extra)}")

        portability = (ROOT / "PORTABILITY.md").read_text(encoding="utf-8")
        inventory_names = TABLE_SKILL_PATTERN.findall(portability)
        if len(inventory_names) != len(set(inventory_names)):
            failures.append("PORTABILITY.md: a skill appears more than once")
        missing_inventory = sorted(skill_names - set(inventory_names))
        extra_inventory = sorted(set(inventory_names) - skill_names)
        if missing_inventory:
            failures.append(
                f"PORTABILITY.md: missing skills: {', '.join(missing_inventory)}"
            )
        if extra_inventory:
            failures.append(
                f"PORTABILITY.md: unknown skills: {', '.join(extra_inventory)}"
            )

    except ValueError as error:
        failures.append(str(error))

    if failures:
        print("Validation failed:", file=sys.stderr)
        for failure in failures:
            print(f"- {failure}", file=sys.stderr)
        return 1

    print(f"Validated {len(skill_dirs)} skills and repository metadata.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/commands/Validate.ts
import { Console, Effect, FileSystem, Path, Schema } from "effect";
import { readImports, trackedSkillPath } from "../imports/metadata.js";

export class ValidationError extends Schema.TaggedError<ValidationError>()(
  "ValidationError",
  { failures: Schema.Array(Schema.String) },
) {}

const allowedFields = new Set([
  "name",
  "description",
  "license",
  "compatibility",
  "metadata",
  "allowed-tools",
]);

const frontmatter = (text: string) => {
  const lines = text.split(/\r?\n/);
  const failures: string[] = [];
  if (lines[0] !== "---")
    return {
      fields: new Map<string, string>(),
      failures: ["missing opening frontmatter delimiter"],
    };
  const end = lines.indexOf("---", 1);
  if (end < 0)
    return {
      fields: new Map<string, string>(),
      failures: ["missing closing frontmatter delimiter"],
    };
  const fields = new Map<string, string>();
  for (const line of lines.slice(1, end)) {
    if (!line || line.trimStart().startsWith("#") || /^\s/.test(line)) continue;
    const match = line.match(/^([a-z][a-z0-9-]*):(?:\s*(.*))?$/);
    if (!match)
      failures.push(
        `invalid top-level frontmatter line: ${JSON.stringify(line)}`,
      );
    else fields.set(match[1] ?? "", match[2] ?? "");
  }
  return { fields, failures };
};

export const validate = Effect.fn("Validate.run")(function* (root: string) {
  const fs = yield* FileSystem.FileSystem;
  const path = yield* Path.Path;
  const failures: string[] = [];
  const entries = yield* fs.readDirectory(root);
  const skillNames: string[] = [];
  for (const entry of entries.sort()) {
    const entryInfo = yield* fs.stat(path.join(root, entry));
    if (entryInfo.type !== "Directory") continue;
    const file = path.join(root, entry, "SKILL.md");
    if (!(yield* fs.exists(file))) continue;
    skillNames.push(entry);
    const text = yield* fs.readFileString(file);
    const parsed = frontmatter(text);
    for (const failure of parsed.failures)
      failures.push(`${entry}: ${failure}`);
    const unknown = [...parsed.fields.keys()].filter(
      (key) => !allowedFields.has(key),
    );
    if (unknown.length > 0)
      failures.push(
        `${entry}: unsupported frontmatter fields: ${unknown.sort().join(", ")}`,
      );
    const name = (parsed.fields.get("name") ?? "").replace(/^['"]|['"]$/g, "");
    if (!name) failures.push(`${entry}: missing non-empty name`);
    else if (!/^[a-z0-9]+(?:-[a-z0-9]+)*$/.test(name))
      failures.push(`${entry}: invalid skill name: ${JSON.stringify(name)}`);
    else if (name !== entry)
      failures.push(
        `${entry}: name ${JSON.stringify(name)} does not match directory ${JSON.stringify(entry)}`,
      );
    if (!parsed.fields.has("description"))
      failures.push(`${entry}: missing description`);
    for (const match of text.matchAll(/(?<!!)\[[^\]]*\]\(([^)]+)\)/g)) {
      const target = (match[1] ?? "").split("#", 1)[0] ?? "";
      if (!target || target.includes("://") || target.startsWith("mailto:"))
        continue;
      if (target.startsWith("/"))
        failures.push(
          `${entry}: absolute Markdown link is not portable: ${target}`,
        );
      else if (!(yield* fs.exists(path.resolve(root, entry, target))))
        failures.push(`${entry}: broken relative Markdown link: ${target}`);
    }
  }
  if (skillNames.length === 0) failures.push("no top-level skills found");
  if (!(yield* fs.exists(path.join(root, "LICENSE"))))
    failures.push("missing repository LICENSE");
  const Catalog = Schema.Struct({
    groupings: Schema.Array(
      Schema.Struct({ skills: Schema.Array(Schema.String) }),
    ),
  });
  const catalogResult = yield* Effect.result(
    fs.readFileString(path.join(root, "skills.sh.json")).pipe(
      Effect.flatMap((text) =>
        Effect.try({
          try: () => JSON.parse(text),
          catch: (cause) => cause,
        }),
      ),
      Effect.flatMap(Schema.decodeUnknownEffect(Catalog)),
    ),
  );
  if (catalogResult._tag === "Failure")
    failures.push(`skills.sh.json: ${String(catalogResult.failure)}`);
  const catalogNames =
    catalogResult._tag === "Success"
      ? catalogResult.success.groupings.flatMap(({ skills }) => skills)
      : [];
  if (new Set(catalogNames).size !== catalogNames.length)
    failures.push("skills.sh.json: a skill appears in more than one grouping");
  const missingCatalog = skillNames.filter(
    (name) => !catalogNames.includes(name),
  );
  const extraCatalog = catalogNames.filter(
    (name) => !skillNames.includes(name),
  );
  if (missingCatalog.length)
    failures.push(
      `skills.sh.json: missing skills: ${missingCatalog.sort().join(", ")}`,
    );
  if (extraCatalog.length)
    failures.push(
      `skills.sh.json: unknown skills: ${extraCatalog.sort().join(", ")}`,
    );
  const portability = yield* fs.readFileString(
    path.join(root, "PORTABILITY.md"),
  );
  const inventory = [...portability.matchAll(/^\| `([^`]+)` \|/gm)].map(
    (match) => match[1] ?? "",
  );
  if (new Set(inventory).size !== inventory.length)
    failures.push("PORTABILITY.md: a skill appears more than once");
  const missingInventory = skillNames.filter(
    (name) => !inventory.includes(name),
  );
  const extraInventory = inventory.filter((name) => !skillNames.includes(name));
  if (missingInventory.length)
    failures.push(
      `PORTABILITY.md: missing skills: ${missingInventory.sort().join(", ")}`,
    );
  if (extraInventory.length)
    failures.push(
      `PORTABILITY.md: unknown skills: ${extraInventory.sort().join(", ")}`,
    );
  const importsResult = yield* Effect.result(readImports(root));
  if (importsResult._tag === "Failure")
    failures.push(`imports.json: ${importsResult.failure.message}`);
  const imports =
    importsResult._tag === "Success"
      ? importsResult.success
      : { version: 1 as const, imports: {} };
  for (const [name, metadata] of Object.entries(imports.imports)) {
    if (
      (metadata.distribution === "wholesale" ||
        metadata.distribution === "official-source") &&
      metadata.localEdits.length !== 0
    )
      failures.push(
        `imports.json: ${name} ${metadata.distribution} import must not declare local edits`,
      );
    if (!metadata.distribution && metadata.localEdits.length === 0)
      failures.push(
        `imports.json: ${name} distributed import must declare local edits`,
      );
    const tracked = trackedSkillPath(root, name, metadata, path);
    if (!(yield* fs.exists(tracked)))
      failures.push(
        `imports.json: missing ${metadata.distribution === "official-source" ? "upstream snapshot" : "skill directory"}: ${name}`,
      );
    if (
      metadata.distribution === "official-source" &&
      (yield* fs.exists(path.join(root, "upstream", name, "SKILL.md")))
    )
      failures.push(`imports.json: ${name} upstream snapshot is discoverable`);
    if (!(yield* fs.exists(tracked))) continue;
    const text = yield* fs.readFileString(tracked);
    const fm = text.split("---", 3)[1] ?? "";
    for (const expected of [
      `# origin: ${metadata.origin}`,
      `# upstream-sha: ${metadata.upstreamSha}`,
      ...(metadata.license ? [`license: ${metadata.license}`] : []),
    ])
      if (!fm.split("\n").includes(expected))
        failures.push(
          `${name}: materialised metadata does not match imports.json`,
        );
    const edits = fm
      .split("\n")
      .filter((line) => line.startsWith("#   - "))
      .map((line) => line.slice(6));
    if (edits.join("\0") !== metadata.localEdits.join("\0"))
      failures.push(
        `${name}: materialised local edits do not match imports.json`,
      );
  }
  for (const name of skillNames) {
    const text = yield* fs.readFileString(path.join(root, name, "SKILL.md"));
    if (
      (text.split("---", 3)[1] ?? "")
        .split("\n")
        .some((line) => line.startsWith("# origin:")) &&
      !imports.imports[name]
    )
      failures.push(`imports.json: missing imports: ${name}`);
  }
  if (failures.length > 0) {
    yield* Console.error(
      `Validation failed:\n${failures.map((failure) => `- ${failure}`).join("\n")}`,
    );
    return yield* new ValidationError({ failures });
  }
  yield* Console.log(
    `Validated ${skillNames.length} skills and repository metadata.`,
  );
});
